Added cusparseSolver, needs GPU to be used

Tongdongq opened the pull request:
This branch contains a linear solver, implemented in cusparse, which can run on an NVIDIA GPU.
To use it, CMake must be able to find CUDA (via find_package(CUDA)). 
The user must pass
```sh
--matrix-add-well-contributions=true --use-gpu=true
```
to the executable in order to use it. If the gpu is selected, but the well contributions are not added, an error is printed and the program exits. Only NORNE is tested, which generates blocks of 3x3. Other blocksizes do not work, but might in the future.

Some notes:
- now exits when `matrix-add-well-contribs` is `false` and `use_gpu` is true, should it print a warning instead and fallback to the cpu? Or maybe an exception can be thrown.
- forward template declarations in `BdaBridge.cpp` are a not very nice
- converting from `BCRSMatrix` to our format and back assumes underlying data is contiguous, testing shows that it is, but is it guaranteed?
- the part from `backend->initialize()` to `backend->solve_system()` should probably be moved to within the backend.
- why `BdaResult`? Because the `cusparseSolverBackend.cu` is compiled using nvcc, adding the `InverseOperatorResult` there causes a compile error:
```
/data/work/tqiu/dune/dune-istl/dune/istl/basearray.hh:101:58: error: ‘typename Dune::Imp::base_array_unmanaged<B, A>::RealIterator’ names ‘template<class B, class A> template<class T> struct Dune::Imp::base_array_unmanaged<B, A>::RealIterator’, which is not a type
       friend class RealIterator<const ValueType>;
```
- timing functions in `cusparseSolverBackend` use `gettimeofday()`
- `BdaBridge` misuses the `std::vector`s a bit.
- if CMake cannot find CUDA, there is no `BdaBridge` created in ISTLSolverEbos.hpp

---

## Results, based on release/2019.04 ##
Intel Xeon E5-2620 v3 @ 2.4GHz, NVIDIA RTX GeForce 2080Ti

### Dune: ###
```
Total time (seconds):         774.473
Solver time (seconds):        767.034
Assembly time (seconds):     298.541 (Failed: 3.21499; 1.0769%)
Linear solve time (seconds): 443.508 (Failed: 5.63787; 1.2712%)
Linear solve setup time (seconds): 14.3272 (Failed: 0.184655; 1.28884%)
Update time (seconds):       11.2785 (Failed: 0.130548; 1.1575%)
Output write time (seconds): 9.36335
Overall Well Iterations:      918 (Failed: 2; 0.217865%)
Overall Linearizations:       1932 (Failed: 21; 1.08696%)
Overall Newton Iterations:    1598 (Failed: 21; 1.31414%)
Overall Linear Iterations:    23982 (Failed: 308; 1.2843%)
```

### cusparseSolver: ###
```
Total time (seconds):         460.21
Solver time (seconds):        452.812
Assembly time (seconds):     293.769 (Failed: 6.40129; 2.17902%)
Linear solve time (seconds): 134.676 (Failed: 4.4542; 3.30735%)
Linear solve setup time (seconds): 14.3 (Failed: 0.37382; 2.61413%)
Update time (seconds):       11.2345 (Failed: 0.259182; 2.30703%)
Output write time (seconds): 9.33874
Overall Well Iterations:      921 (Failed: 4; 0.434311%)
Overall Linearizations:       1940 (Failed: 42; 2.16495%)
Overall Newton Iterations:    1605 (Failed: 42; 2.61682%)
Overall Linear Iterations:    24387 (Failed: 861; 3.53057%)
```

blattms commented:
jenkins build this please

blattms commented:
Currently testing this a bit. It seems like on my system the difference between CPU and GPU is quite different, due to my more consumer friendly graphics card (300 EUR instead of 750-1200 EUR)

CPU: AMD Ryzen Threadripper 1950x 16 cores 3.4 GHz
GPU: GeForce GTX 1060 6 GB

I have used default settings for everything but `--use-gpu` and `--matrix-add-well-contributions`.

| Model |  GPU  | Serial CPU well-contrib=true | Serial CPU well-contrib=false |
| --- | ---  | --- | --- |
| Norne | 518s |  575s | 600s |

I also tried to run model 2, but the `--matrix-add-well-contributions=true` option results in time step chopping until the simulator aborts (For the CPU as well as the GPU).

Tongdongq commented:
The Intel Xeon E5-2620 v3 is from 2014 and runs at 2.4 GHz. This might account for your faster CPU benchmark.
The GTX 1060 6 GB has 1280 CUDA cores and 192 GB/s bandwidth, the RTX 2080Ti has 4352, and about 616 GB/s bandwidth.
Do you have the time split into assembly and solving? This will make comparing the two GPUs easier.
Does the cusparseSolver converge with model2 systems?

alfbr commented:
> t seems like on my system the difference between CPU and GPU is quite different, due to my more consumer friendly graphics card

Cool, thanks for sharing! That means we even in this case get a small speed-up. This is a very good start for the GPU path, thanks for the good work @Tongdongq :-)

blattms commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,249 @@
+	int nnz = mat.nonzeroes()*dim*dim;
+
+	// copy nonzeros
+	memcpy(h_vals.data(), &(mat[0][0][0][0]), sizeof(double)*nnz);
```
I think this currently works because we one of the ways to set up BCRSMatrix such that the data is in a contiguous array.
I do not think that this will be changed, but a comment about it should not hurt.

If the memory is continues, the shouldn't we skip the copying and just pass `reinterpret_cast<double>(&(mat[0][0][0][0])`
to all the cuda calls that copy memory? Or is there a reason that I do not see?

Tongdongq replied:
No reason, I think this extra copy could be skipped, like done in `get_result`. I'll try it out.

blattms commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,249 @@
+// converts the BlockVector b to a flat array
+template <class BridgeVector>
+void convertBlockVectorToArray(BridgeVector& b, std::vector<double> &h_b) {
+	memcpy(h_b.data(), &(b[0]), sizeof(double) * b.N() * b[0].dim());
```
Here too: Maybe we should omit the copying and use a reinterpret cast?

blattms commented:
@alfbr `sed s/consumer/media crazy gamer/`. Personally, I would never have bought that card without the need for it in a paid project.

blattms commented:
@Tongdongq It might th different memory bandwidth of the GPUs.

Complete timing:

|      | GPU | CPU |
| --- | ---      | ---    |
| Total time (seconds):         | 517.852 | 574.731 |
| Solver time (seconds):      |  517.779 | 574.661 |
| Assembly time (seconds): |    251.44 | 243.055 |
| Linear solve time (seconds): | 224.019 | 289.172 |
| Linear solve setup time (seconds): | 12.5784 | 12.3379 |
| Update time (seconds): |      11.5952 | 11.2752 |
| Output write time (seconds): | 10.8919 |11.3409 |

Tongdongq commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -435,11 +465,26 @@ class WellModelMatrixAdapter : public Dune::AssembledLinearOperator<M,X,Y>
+                }else{
+                    // CPU fallback, or default case for Dune
+                    auto precond = constructPrecond(linearOperator, parallelInformation_arg);
+                    solve(linearOperator, x, istlb, *sp, *precond, result);
```
The cusparseSolver prints a warning when the preconditioner could not be created.
If the cusparseSolver fails to converge within `maxit` iterations, Dune might also not converge, given that they have the same settings.
I added a print. I can also print the number of fallbacks at the end, but the only place I know is the destructor. If it needs to be printed somewhere else, you'll have to tell me where.

Tongdongq commented on opm/simulators/linalg/bda/cusparseSolverBackend.cu:
```diff
@@ -0,0 +1,454 @@
+		cudaMemcpyAsync(d_bRows, rows, (Nb+1) * sizeof(int), cudaMemcpyHostToDevice, stream);
+		cudaMemcpyAsync(d_b, b, N * sizeof(double), cudaMemcpyHostToDevice, stream);
+		cudaMemsetAsync(d_x, 0, sizeof(double) * N, stream);
+
```
I don't think this is explicitly required. All GPU functions of one cusparseSolver instance use the same cudaStream. Functions within the same stream are not parallelized. Using cudaStreams is meant to allow running multiple independent instances of Flow on the same GPU.
I only added them in the `verbosity` blocks now, so timing can be done accurately.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,249 @@
+			backend->update_system_on_gpu(h_vals.data(), h_b.data());
+		}
+		backend->reset_prec_on_gpu();
+		if(backend->create_preconditioner()){
```
The function `cusparseXbsrilu02_zeroPivot` contains an implicit synchronization point, as it calls `cudaDeviceSynchronize`. But since `cusparseDbsrilu02` is a cusparse function, and the stream from the cusparseSolver is attached to the cusparse library, it should wait for the memcpys to be done.
There is a warning inside `create_preconditioner` if the preconditioner cannot be created.

Tongdongq replied:
cusparseSolver now returns a cusparseSolverStatus value, which is interpreted by BdaBridge. The Bridge prints a warning using OpmLog.

alfbr commented:
> It might th different memory bandwidth of the GPUs.

Sounds plausible. It will be interesting to see how these numbers look with multiple realizations running simultaneously (i.e., saturating the bandwidth of both CPU and GPU).

blattms commented:
The problem with warnings in *.cu files is that they are using `fprint` and cannot use the OpmLog facility. I would recommend to move the warnings one level up and use `OpmLog::warning`.

blattms commented:
I forgot one thing: Please remove the compiler warnings generated, see https://ci.opm-project.org/job/opm-simulators-PR-builder/447/gcc/new/

blattms commented:
I will misuse this PR to force a parallel norne run of the parallel implementation of master on jenkins to compare with #2205

blattms commented:
jenkins run norne_parallel

blattms commented:
Forgot the magic word. Should be more polite to butlers.

blattms commented:
jenkins run norne_parallel please

Tongdongq commented:
@blattms I addressed some of your comments.
Notes:
- BdaBridge still has `printf`, these might need to be changed to `OpmLog`
- cusparseSolver.cu also has `printf`, these can be changed to `OpmLog` as well

alfbr commented:
Yes, please use the OpmLog mechanism to output information in the terminal window. The only complexity is that you need to categorize them. The categories "info", "warning" and "error" are hopefully self-explanatory.

Tongdongq commented:
I changed the `printf` to `ostringstream`, but this uses 3 lines instead of 1 when printing variables.
If someone tries to use cusparseSolver with an unsupported blocksize, OpmLog prints an error, the Bridge then disables the gpu, and Dune is used for the rest of the simulation. We could change this to actually exiting.
The Bridge still contains a `PRINT_TIMERS_BRIDGE` macro. This could be changed to a runtime variable, merged with `flow-linear-solver-verbosity` or removed altogether. The Bridge prints are not that interesting anyway.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,242 @@
+			break;
+		default:
+			OpmLog::warning("cusparseSolver returned unknown status code");
+		}
```
BdaBridge now receives a cusparseSolverStatus and prints a warning if something is wrong.

Tongdongq commented on CMakeLists.txt:
```diff
@@ -282,3 +292,12 @@ endif()
+
+# must link libraries after target 'flow' has been defined
+if(CUDA_FOUND)
+  target_link_libraries( flow ${CUDA_cublas_LIBRARY} )
```
I don't think it works like that. I got an error during linking:
``
/path/ld: lib/libopmsimulators.a(cusparseSolverBackend.cu.o): undefined reference to symbol 'cublasCreate_v2@@libcublas.so.10'
//lib64/libcublas.so.10: error adding symbols: DSO missing from command line
``

blattms commented:
You need to use the correct order. symbols will only be search in libraries specified afterwards.
Hence first comes the cupsparse library and then the cublas library.

Tongdongq commented:
> You need to use the correct order. symbols will only be search in libraries specified afterwards. Hence first comes the cupsparse library and then the cublas library.

I cannot reproduce this. I still get linker errors whenever I don't add both libraries (cublas/cusparse) to both targets (flow/ebos). If I reverse the order of linking, it still works.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,242 @@
+				if(r.index() == c.index()){
+					for(int rr = 0; rr < dim; ++rr){
+						// pointer arithmetic
+						int offset = (int)((long unsigned)&(mat[r.index()][c.index()][rr][rr]) - (long unsigned)nnzs); // in bytes
```
We also rely on the memory layout in `static_cast<double*>(&(((*mat)[0][0][0][0])))`.
I propose
```
    for(typename BridgeMatrix::size_type i = 0; i < mat.N(); ++i){
    	for(typename BridgeMatrix::size_type j = 0; j < dim; ++j){
    		double val = mat[i][i][j][j];
    		if(val == 0.0){ // could be replaced by '< 1e-30' or similar
    			mat[i][i][j][j] = zero_replace;
    			++numZeros;
    		}
    	}
    }
```
This removes the std::vector diag_indices. mat[i][i] is pretty much guaranteed to exist. For every row, it gets the block on the diagonal (mat[i][i]), and in that block, it checks the values on the diagonal (mat[i][i][j][j]).

blattms replied:
You kind of have point. But if we rely on something we might want to check that it holds. e.g.
`&((mat.begin()+1).begin()[0][0]) - (&mat.begin().begin()[0][0]) == mat.begin()->size()*blocksize*blocksize;`

Your proposal triggers at least three binary searches per matrix row. I thought the `diag_indices` were there to prevent this.

Tongdongq replied:
I see. How does your offset work, and how is it used? I can't really follow it.

blattms replied:
`offset` is the offset of `mat[i][i][j][j]`. `roff` is the offset of the current block row in your cuda matrix values array. `c-r.begin()` is the offset of the nonzero counted from the start of the in block. Then there is a mistake. In the rest please substitute `r` with `rr`:
``offset = roff + (c - r.begin())+ rr * dim + rr;``
The inner loop (over the rows in the matrix block) is omitted. Its loop body is the culculation of `offset`. The last line is in the outer loop.

Tongdongq replied:
I'm getting the same address for `mat[r.index()][c.index()].begin()` and `mat[r.index()].begin()`, but they are different from `&(mat[r.index()][c.index()][rr][rr])`.

blattms replied:
Not sure I get where you are trying to go with that statement. Seems like we might be talking about different things.

>
> I'm getting the same address for `mat[r.index()][c.index()].begin()` and `mat[r.index()].begin()`

That can be the	case, but only for the very first non-zero of a	row.
```
if(r.size()>1){
  auto nextC = ++(r.begin());
  assert(&(*mat[r.index()][nextC.index()].begin())!=&(*mat[r.index()].begin()))
}
```
should never abort

> , but they are different from `&(mat[r.index()][c.index()][rr][rr])`.

Only for the case that `rr==0` holds,
`assert(&(mat[r.index()][c.index()][rr][rr])==&(*mat[r.index()][c.index()].begin())`
should not abort.

What I was trying to say is:
No matter what blocksize we have, `c - r.begin()` will tell the consecutive zerostatring index of the current column. Let us say that for the current row, we have a nonzero blocks with column indices `[0, 3, 4, 5]`. If c is positioned at column index `4`, then `c - r.begin()==2` will hold. Does this help.
`c-rbegin

Tongdongq replied:
The offset (in number of blocks) between the first block of the row and the block that we are interested in (diagonal block) is `int offset = mat[r.index()].find(r.index()).offset())`. This is what you mean by `c - r.begin()`.
I can perform `double *valp = mat[r.index()].getptr() + offset` to get the address of the first value in the block, and then use `valp += dim + 1` to get the 2nd diagonal value of the block.
But it uses some function that are not meant to be used outside of dune-istl.

I think we should only check if the data is contiguous, and use the CPU otherwise. If we have to copy the nonzeroes per row, we'll have 2x44321 `memcpy`s for NORNE. I suspect this defeats the purpose of the GPU.

Tongdongq replied:
It seems to work using these complicated expressions:
```
for every row:
size_type offset = mat[r.index()].find(r.index()).offset();
for(int rr = 0; rr < dim; ++rr){
    double val = (*(mat[r.index()].getptr() + offset))[rr][rr];
    if(val == 0.0){
        (*(mat[r.index()].getptr() + offset))[rr][rr] = zero_replace;
        ++numZeros;
    }
    diag_indices.emplace_back(offset);
}
```
This stores an offset per bsr row, instead of every csr row. But it is rather ugly.

blattms replied:
That would do. For beauty, shouldn't this also work:
```
auto diag = r->find(r.index());
assert(diag.index() == r.index());
for(int rr = 0; rr < dim; ++rr){
   auto& val = diag[rr][rr]; // reference to easily change the value
   if (val ==0.0){
      val = zero_replace;
      ++numZeros;
     }
    diag_indices.emplace_back(diag.offset());
}
```
I must say that I do not understand the original offset code any further (even less the more often I look at it). There is another line below these in the original version that does
```
					offset /= sizeof(double);  // convert offset to doubles
```
before storing it. There I am totally lost. Why divide the offset by `8`?

Tongdongq replied:
For the original offset code:

1. Set `double *nnzs = &(mat[0][0][0][0])`. It points to the start of the underlying array with nonzeroes, which is assumed to be contiguous.
2. For row `r`, and row `rr` in the block, find the address of `mat[r][r][rr][rr]`.
3. Calculate the number of bytes between `nnzs` and the value. Divide by 8 to get the offset in number of doubles.
4. To use the offsets: for every offset, check `nnzs[offset]`. `nnzs` is a `double*` so the index must indicate doubles.

I will try your snippet.

Tongdongq replied:
I got it working, storing:
```c++
auto diag = r->find(r.index());  // diag is an iterator
assert(diag.index() == r.index());
for(int rr = 0; rr < dim; ++rr){
    auto& val = (*diag)[rr][rr]; // reference to easily change the value
    if (val == 0.0){
        val = zero_replace;
        ++numZeros;
    }
}
diag_indices.emplace_back(diag.offset());
```
Using:
```c++
for(typename BridgeMatrix::iterator r = mat.begin(); r != mat.end(); ++r){
    typename BridgeMatrix::size_type offset = diag_indices[r.index()];
    auto& diag_block = r->getptr()[offset]; // diag_block is a reference to MatrixBlock, located on column r of row r
    for(int rr = 0; rr < dim; ++rr){
        auto& val = diag_block[rr][rr];
        if(val == 0.0){
            val = zero_replace;
            ++numZeros;
        }
    }
}
```
But it still uses the `r->getptr()` construct. I think the only alternative is to use `find()`, which performs the binary search.

blattms replied:
I would have assumed that `r->begin()+offset` does the same as `r->getptr()[offset]`. Anyway, both look good and it seems a matter of test. You do not need to invest more time here. Would you please push your changes (including the replacement of tabs with space and give me a heads up. I guess this is good to be merged.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,242 @@
+#if HAVE_CUDA
+template <class BridgeMatrix>
+int checkZeroDiagonal(BridgeMatrix& mat) {
+	static std::vector<int> diag_indices;   // contains offsets of the diagonal nnzs
```
We might not need this, using the approach below.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,242 @@
+				h_rows.emplace_back(sum_nnzs);
+			}
+		// set last rowpointer
+		h_rows[mat.N()] = mat.nonzeroes();
```
`mat.nonzeroes()` only triggers a `std::accumulate` when it wasn't already calculated. The comment in `bcrsmatrix.hh` suggests this only happens when using row-wise allocation.
The last value in h_rows should indeed already be correct. Do you prefer `assert` over `OPM_THROW` in this case?

blattms replied:
I guess `OPM_THROW` is better in this case as this is not performance critial.

blattms commented:
Sorry, my bad. Did not see that these are the two binaries.  How about:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index fd1e2ed28..0e42f10cd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -299,9 +299,7 @@ endif()
 
 # must link libraries after target 'flow' has been defined
 if(CUDA_FOUND)
-  target_link_libraries( flow ${CUDA_cublas_LIBRARY} )
-  target_link_libraries( flow ${CUDA_cusparse_LIBRARY} )
-  target_link_libraries( ebos ${CUDA_cublas_LIBRARY} )
-  target_link_libraries( ebos ${CUDA_cusparse_LIBRARY} )
+  target_link_libraries( opmsimulators ${CUDA_cusparse_LIBRARY} )
+  target_link_libraries( opmsimulators ${CUDA_cublas_LIBRARY} )
 endif()
```

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,242 @@
+
+		if(dim != 3){
+			OpmLog::error("cusparseSolver only accepts blocksize = 3 at this time");
+			use_gpu = false;
```
I changed it to
`OpmLog::warning("cusparseSolver only accepts blocksize = 3 at this time, will use Dune for the remainder of the program");`

blattms commented:
Two more things:
- I noticed when rebasing: In the added files you are using tabs for the indentation, please use [4 spaces](https://wiki.opm-project.org/index.php?title=Suggested_coding_standard#whitespace)
- In some parts of the code you are checking that the blocksize is 3. Is there a special reason for that? It seems like most of the code should work for other block sizes.

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -0,0 +1,242 @@
+		static std::vector<int> h_cols;
+		int dim = (*mat)[0][0].N();
+		int N = mat->N()*dim;
+		int nnz = mat->nonzeroes()*dim*dim;
```
>     * In some parts of the code you are checking that the blocksize is 3. Is there a special reason for that? It seems like most of the code should work for other block sizes.
We have only checked for blocksize 3, other sizes should indeed work. But we do not have testcases for that (or we don't know we have them, spe-models?).

Tongdongq commented:
>     * In some parts of the code you are checking that the blocksize is 3. Is there a special reason for that? It seems like most of the code should work for other block sizes.

We have only checked for blocksize 3, other sizes should indeed work. But we do not have testcases for that (or we don't know we have them, spe-models?).

bska commented:
jenkins build this please

blattms commented on CMakeLists_files.cmake:
```diff
@@ -129,6 +134,10 @@ list (APPEND PUBLIC_HEADER_FILES
   opm/simulators/aquifers/BlackoilAquiferModel.hpp
   opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp
+  opm/simulators/linalg/BdaBridge.hpp
+  opm/simulators/linalg/BdaResult.hpp
```
jenkins says that these headers are not there. Maybe they reside in `opm/simulators/linalg/bda`?

Tongdongq replied:
Yes, I fixed the paths now.

blattms commented:
jenkins build this please

bska commented:
jenkins build this please

blattms commented:
In it goes. thanks a lot for the contribution.

atgeirr commented:
I second that, thanks for your work @Tongdongq. Also to @blattms for doing a lot of work reviewing this.
